protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+
+constructor(Addresses _implementations, ISystemConfigs[] _systemConfigs) {
+  implementations = _implementations;
+  for(uint i=0; i < _systemConfigs.length; i++){
+    systemConfigs.push(_systemConfigs[i])
+  }
+}
+
+function upgrade() public {
+  for(uint i=0; i< systemConfigs.length; i++) {
+    // Read the `Addresses` struct from each `SystemConfig` in the `systemConfigs` mapping.
+    // For each entry in the `Addresses` struct:
+      // 1. Call the `SuperchainProxyAdmin` to update the implementation to the `Populator` contract and call `Populator.upgrade()`.
+      // 2. Update the implementation to the final implementation contract.
+  }
+  // Return ownership of the `SuperchainProxyAdmin` to the Upgrade Controller (`msg.sender`)
+      //    `SuperchainProxyAdmin`. The unique upgrade call encoding required for the non-standard proxies
+      //    should be handled here.
+}
+```
+
+**Note:**A new getter should be added to the `SystemConfig` to perform this in a single call.
+
+## L1ChugsplashProxy Considerations
+
+The L1 Chugsplash proxy is much more expensive to upgrade, as it is done by calling
+[`setCode()`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/legacy/L1ChugSplashProxy.sol#L92)
+which creates a new implementation contract. In this sense the L1StandardBridge is not truly MCP-L1
+ready.
+
+This will required an intermediate proxy to enable pointing to a single implementation contract,
+options for this intermediate proxy include:
+
+1. **Cheaper upgrades, costlier bridging:** a onetime upgrade which sets the implementation for all L1
+   Standard bridges to an intermediate standard Proxy contract. From then on the standard proxy
+   will have its implementation upgraded.
+   This is a tradeoff between token bridging costs and upgrade
+   costs. It may also add complexity in the form of needing to track a new contract for each OP
+   Chain.
+2. **Middle ground:** each time the implementation is changed, `setCode()` could deploy a minimal
+   intermediate immutable proxy which hardcodes the address of the implementation. This reduces the
+   size of the contract, is more gas efficient (no `sload` in the intermediate proxy) and avoids the
+   need to track the intermediate proxies.
+3. **Costlier upgrades, cheaper bridging:** deploy a new but identical implementation for every
+   L1 Standard Bridge. Off chain analysis can easily verify that the implementatin matches the
+   governance approved bytecode.
+
+Further gas cost analysis and product discussion will be required to make a final decision.
+
+## Development and Testing considerations
+
+This solution should be implemented such that it both enables and _requires_ developers to implement
+the upgrade path from the most recent release to the one curently being developed. This means it
+will be necessary to have a deployment of the previous release available on the current commit, in
+order to test the upgrade path.
```

Comment:
This will probably not work for proofs in the short-medium term.

We have `kona` + `asterisc` shipping soon (likely before Ithsmus on OP mainnet), which will be the first of several alternative proof systems deployed on the superchain. `asterisc` for example has a dependency on the monorepo, and has its FPVM contract (`RISCV.sol`) in that repository. We keep this separation to not bloat the monorepo, as the monorepo's contracts provide the interface for these alternative proof systems. Once we start moving towards multi-proofs, we'll likely have more examples of this, for example for a dispute game backed by a ZKP.

It would be great if OPCM can support contracts that do not exist within the monorepo to support the extensibility of the proof system (and other parts of the protocol) without requiring that everything must be upstreamed into the monorepo.

Reply:
If these are becoming part of the core protocol, it seems they should live in the monorepo? Or could the dependency graph be flipped so the monorepo pulls them in as a dependency? Having them in the monorepo feels significantly simpler:
- Fits into the OPCM upgrades process
- All of our existing solidity and CI checks get run for free
- Easy to add new e2e tests with these proof systems in the monorepo
- No additional changes needed to our, development, release, and governance processes

Having key parts of the protocol in a separate repo feels like it opens a lot of questions, what is the plan for addressing all of the above?

Reply:
The decision was made to vendor the `RISCV.sol` contract and this has been done in https://github.com/ethereum-optimism/optimism/pull/12994, so I will resolve this thread